# Customer Churn Analysis
    The "Churn_Modelling.csv" dataset contains information about customers of a bank. 
    Problem Statement: The bank has been seeing unusual churn rate and they want to assess and address the problem.
    
    The bank has released information related to 10,000 customers from European region.
    GOAL: our job is to create a GeoDemographic Segmentation Model to help the bank identify which customers are at the highest risk of leaving. 
    
    *Geo Demographic Segmentation have lots of application in determining decision making based on Geo Demographic Scenarios.
    
    Observation: The outcome is binary in nature.
    
    


In [87]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns




### Looking at data. Understanding the structure.

In [88]:
data = pd.read_csv('Churn_Modelling.csv')
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [89]:
data.describe()

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

In [90]:
X = data.iloc[:,3:13].values
Y = data.iloc[:,13].values

In [91]:
type(X)

<class 'numpy.ndarray'>

In [92]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder

In [93]:
labelEncoder_X_1 = LabelEncoder()
X[:,1] = labelEncoder_X_1.fit_transform(X[:,1])
labelEncoder_X_2 = LabelEncoder()
X[:,2] = labelEncoder_X_2.fit_transform(X[:,2])

In [94]:
X[0]

array([619, 0, 0, 42, 2, 0.0, 1, 1, 1, 101348.88], dtype=object)

In [95]:
oneHotEncoder = OneHotEncoder(categorical_features=[1])

In [96]:
X = oneHotEncoder.fit_transform(X).toarray()

In [98]:
X = X[:,1:]
len(X[0])

11

## Splitting Dataset

In [99]:
from sklearn.model_selection import train_test_split

In [100]:
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2,random_state = 0)

In [101]:
len(X_test),len(X_train)

(2000, 8000)

# Feature Scaling 
    Feature Scaling of X_train and X_test(REQUIRED to ease on the large amount of computations)
    Artificial Neural Networks are very compute intensive. Feature scaling makes it easy doing the computations

In [102]:
from sklearn.preprocessing import StandardScaler

In [103]:
sc = StandardScaler()

In [104]:
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# Part 2 - Making the ANN using the data
    Will require Keras Library and some modules within it

In [105]:
import keras

### We need 2 modules within Keras
    1> Sequential Module - To Initialize the ANN
    2> Dense Module - To build layer of ANN

In [106]:
from keras.models import Sequential
from keras.layers import Dense

    We initialize the Deep learning model by defining it as a sequence of Layers

### Designing the Artificial Neural Network

In [107]:
# Initializing the ANN
classifier = Sequential()

# Adding the Input Layers and the first hidden Layer
"""
Adding number of layers is said to be an art. It requires experimentation

But a goot TIP is to:
No. of Nodes in the Hidden layer = Average(No. of Nodes in InputLayer,No. of Nodes in OutputLayer) 
x  = (11 + 1)/2 = 6 Nodes

Input Dimensions = 11
We need to specify it in the Dense() for the First Hidden Layer


Otherwise,
We can use K-fold CrossValidation to determine it
"""

# Adding the input layer and the first hidden layer (Input Dim required for First Hidden Layer)
# InputLayer + Hidden Layer 1
classifier.add(Dense(output_dim = 6,init = "uniform",activation="relu",input_dim = 11))

# Hidden Layer 2 
classifier.add(Dense(output_dim = 6,init = "uniform",activation = "relu"))

# Output Layer (Activation="sigmoid" and output_dim = 1,  because O/p is Binary) 

# If dependent variable has more than 2 catergories, Use SOFTMAX 
classifier.add(Dense(output_dim = 1,init = "uniform",activation = "sigmoid"))


/Users/Rahul/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:22: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=11, units=6, kernel_initializer="uniform")`
/Users/Rahul/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:25: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", units=6, kernel_initializer="uniform")`
/Users/Rahul/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:30: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="sigmoid", units=1, kernel_initializer="uniform")`


## Compiling the ANN
    Now that we have added all the layers, we move towards compiling
    Applying Stochastic Gradient Descent for the ANN

In [108]:
help(classifier.compile)

Help on method compile in module keras.engine.training:

compile(optimizer, loss=None, metrics=None, loss_weights=None, sample_weight_mode=None, weighted_metrics=None, target_tensors=None, **kwargs) method of keras.engine.sequential.Sequential instance
    Configures the model for training.
    
    # Arguments
        optimizer: String (name of optimizer) or optimizer instance.
            See [optimizers](/optimizers).
        loss: String (name of objective function) or objective function.
            See [losses](/losses).
            If the model has multiple outputs, you can use a different loss
            on each output by passing a dictionary or a list of losses.
            The loss value that will be minimized by the model
            will then be the sum of all individual losses.
        metrics: List of metrics to be evaluated by the model
            during training and testing.
            Typically you will use `metrics=['accuracy']`.
            To specify different me

In [109]:
# A very efficient type of the Stochastic Gradient Algo is called "adam"
# Loss function for Output Sigmoid 
#       Loss for Binary Output = binary_crossentropy
#       Loss for Categorical Output = Categorical_CrossEntropy
classifier.compile(optimizer="adam",loss="binary_crossentropy",metrics=["accuracy"])

### Fitting the ANN to the Training Set

In [110]:
X_train
Y_train

array([0, 0, 0, ..., 0, 0, 1])

In [111]:
# "BatchSize" is the number of observations after which we want to update the weights
# BatchSize = 10 (For Stochastic Gradient Descent, it is small batches for which the weights are adjusted)
# Epoch -> 1 full interation of all observations over the whole ANN
classifier.fit( X_train , Y_train , batch_size=1 ,epochs=100 )

Epoch 1/100
8000/8000 [==============================] - 8s 1ms/step - loss: 0.4427 - acc: 0.7959
Epoch 2/100
8000/8000 [==============================] - 7s 899us/step - loss: 0.4210 - acc: 0.8188
Epoch 3/100
8000/8000 [==============================] - 7s 895us/step - loss: 0.4155 - acc: 0.8294
Epoch 4/100
8000/8000 [==============================] - 7s 904us/step - loss: 0.4125 - acc: 0.8324
Epoch 5/100
8000/8000 [==============================] - 7s 895us/step - loss: 0.4109 - acc: 0.8325
Epoch 6/100
8000/8000 [==============================] - 7s 915us/step - loss: 0.4089 - acc: 0.8337
Epoch 7/100
8000/8000 [==============================] - 8s 966us/step - loss: 0.4085 - acc: 0.8307
Epoch 8/100
8000/8000 [==============================] - 8s 1ms/step - loss: 0.4081 - acc: 0.8347A: 0s - loss: 0.4087 - acc:
Epoch 9/100
8000/8000 [==============================] - 7s 911us/step - loss: 0.4068 - acc: 0.8353
Epoch 10/100
8000/8000 [==============================] - 7s 928us/step - los

8000/8000 [==============================] - 15s 2ms/step - loss: 0.4023 - acc: 0.8353: 2s -  - ETA: 1s - loss: 0.405 - ETA: 0s - loss: 0.4030 - acc: 0.8
Epoch 77/100
8000/8000 [==============================] - 16s 2ms/step - loss: 0.4021 - acc: 0.8346
Epoch 78/100
8000/8000 [==============================] - 17s 2ms/step - loss: 0.4022 - acc: 0.8334: 4s - los - ETA: 3s - loss: 0.3972 - acc: 0. - ETA: 3s - los
Epoch 79/100
8000/8000 [==============================] - 15s 2ms/step - loss: 0.4024 - acc: 0.8332: 7s - loss:
Epoch 80/100
8000/8000 [==============================] - 8s 1ms/step - loss: 0.4022 - acc: 0.8314
Epoch 81/100
8000/8000 [==============================] - 599s 75ms/step - loss: 0.4024 - acc: 0.8349
Epoch 82/100
8000/8000 [==============================] - 27s 3ms/step - loss: 0.4024 - acc: 0.8339
Epoch 83/100
8000/8000 [==============================] - 17s 2ms/step - loss: 0.4024 - acc: 0.8339
Epoch 84/100
8000/8000 [==============================] - 18s 2ms/step -

## Predicting the Test Set results
    Although we have obtained the accuracy after each Epoch and we know the accuracy of the predictions on the Training set, We need to see how the classifier performs on the Test Set

In [116]:
Y_pred = classifier.predict(X_test)

In [117]:
Y_pred

array([[0.17325348],
       [0.3658588 ],
       [0.19470961],
       ...,
       [0.21627688],
       [0.1456406 ],
       [0.12692833]], dtype=float32)

## Making the Confusion Matrix

In [118]:
from sklearn.metrics import confusion_matrix

In [121]:
for i,pred in enumerate(Y_pred):
        Y_pred[i] = 1 if pred >= 0.5 else 0
        
cm = confusion_matrix(Y_test,Y_pred)

In [122]:
cm

array([[1533,   62],
       [ 252,  153]])

In [125]:
accuracy = (cm[0][0]+cm[1][1])/(cm[0][0]+cm[1][0]+cm[0][1]+cm[1][1])
accuracy

0.843

# We got an accuracy of 84.3 %
    accuracy = number of correct predictions/sum of all prediction